File: head_first_design_pattern/Python/decorator.py
from abc import ABC, abstractmethod


class Beverage(ABC):
    def __init__(self):
        self.description = "Unknown Beverage"

    def get_description(self):
        return self.description

    @abstractmethod
    def cost(self):
        pass


class CondimentDecorator(Beverage):
    @abstractmethod
    def get_description(self):
        pass


class Espresso(Beverage):
    def __init__(self):
        self.description = "Espresso"

    def cost(self):
        return 1.99


class HouseBlend(Beverage):
    def __int__(self):
        self.description = "House Blend Coffee"

    def cost(self):
        return 0.89


class Mocha(CondimentDecorator):
    def __init__(self, beverage):
        self.beverage = beverage

    def get_description(self):
        return self.beverage.get_description()+", Mocha"

    def cost(self):
        return 0.20 + self.beverage.cost()


beverage = Espresso()
print("{}${}".format(beverage.get_description(), beverage.cost()))

beverage2 = HouseBlend()
beverage2 = Mocha(beverage2)
beverage2 = Mocha(beverage2)
print("{}${}".format(beverage2.get_description(), beverage2.cost()))

File: head_first_design_pattern/C++/decorator.cpp
#include<iostream>
#include<string>

using namespace std;

class Beverage {
public:
	string description;  
	Beverage() {
		this->description = "Unkown Beverage";
	};

	string get_description() {
		return this->description;
	}

	virtual float cost()=0;
};

class CondimentDecorator :public Beverage {
public:
	CondimentDecorator() {};
	virtual string get_description()=0;
};


class Espresso :public Beverage {
public:
	Espresso() { this->description = "Espresso";};

	float cost(){
		return 1.99;
	}
};

class HouseBlend :public Beverage {
public:
	HouseBlend() { this->description = "House Blend Coffee"; };
	
	float cost() {
		return 0.99;
	}
};

class Mocha :public CondimentDecorator{
public:
	Beverage *beverage;
	Mocha(Beverage *beverage) {
		this->beverage = beverage;
	};

	string get_description() {
		return this->beverage->get_description() + ", Mocha";
	}

	float cost() {
		return this->beverage->cost() + 0.20;
	}
};

int main(void) {
	Beverage *beverage = new Espresso();
	cout << beverage->get_description()<<"$"<<beverage->cost() << endl;

	//not worked, beverage2 call base function.
	Beverage *beverage2 = new HouseBlend();
	beverage2 = new Mocha(beverage2);
	cout << beverage2->get_description() << "$" << beverage2->cost() << endl;
	/*beverage2 = new Mocha(beverage2);
	cout << beverage2->get_description()<<"$"<<beverage2->cost()<<endl;*/
	while (1);
	return 0;
}